nx/lib/nx/defn/grad.ex

```diff
@@ -424,6 +424,90 @@ defmodule Nx.Defn.Grad do
     to_grad(x, g, cache)
   end
 
+  defp grad(:concatenate, [tensors, axis], %{shape: ans_shape} = ans, g, cache) do
+    # temporary code annotations -- should not merge with this
+    """
+    # operand_shapes = Enum.map(tensors, & &1.shape)
+    -- operand_shapes = [o.aval.shape if ad.is_undefined_primal(o) else o.shape
+    --                  for o in operands]
+
+    # cumulative sum [1,2,3,4] -> [1,3,6,10]
+    --limit_points = np.cumsum([shape[dimension] for shape in operand_shapes])
+
+    # starts.shape = length(tensors) x tuple_size(Enum.at(operand_shapes, 0))
+    # starts = [
+    # [0,0,... 0]
+    # [0,0, ..., limit_points[1], 0, 0]
+    # [0,0, ..., limit_points[2], 0, 0]
+    # ...
+    # ]
+    --starts = np.zeros((len(operands), t.ndim), dtype=int)
+    --starts[1:, dimension] = limit_points[:-1]
+
+    # limits = List.duplicate(ans.shape, length(tensors)) |> Enum.zip(limit_points) |> Enum.map(fn {row, p} -> List.replace_at(row, axis, p) end)
+    -- limits = np.tile(t.shape, (len(operands), 1))
+    -- limits[:, dimension] = limit_points
+
+
+
+    # slice here is slice(tensor, start_indices, end_indices), instead
+    # of Nx.slice which is (tensor, start_indices, lenghts)
+    return [slice(t, start, limit) if ad.is_undefined_primal(o) else None
+        for o, start, limit in zip(operands, starts, limits)]
+    """
+
+    operand_shapes = Enum.map(tensors, & &1.shape)
+    IO.inspect(operand_shapes, label: "operand shapes")
+
+    limit_points =
+      Enum.reduce(operand_shapes, [], fn
+        x, [] -> [elem(x, axis)]
+        x, [h | _] = acc -> [elem(x, axis) + h | acc]
+      end)
+      |> Enum.reverse()
```

Comment:
```suggestion
    limit_points = Enum.scan(tensors, 0, fn t, acc -> elem(t.shape, axis) + acc end)
```

Or if you want to be super concise:

```suggestion
    limit_points = Enum.scan(tensors, 0, &elem(&1.shape, axis) + &2)
```